### PR TITLE
Use presentation compiler for REPL tab completion

### PR DESCRIPTION
**Pull request by retronym**
This brings tab-completion in the REPL in line with the experience in IDEs and ENSIME-powered editors.

* Reliable completion, also in partial expressions and syntactically incorrect programs: try `class C { def f(l: List[Int]) = l.<TAB>`
* CamelCase completion: try `(l: List[Int]).rro<TAB>`, it expands to `(l: List[Int]).reduceRightOption`
* Find members by typing any CamelCased part of the name: try `classOf[String].typ<TAB>` to get `getAnnotationsByType`, `getComponentType` and others
* Complete non-qualified names, including types: try `def f(s: Str<TAB>`
* Show desugarings performed by the compiler by adding `//print`: try `for (x <- 1 to 10) println(x) //print<TAB>`
* Press tab twice to see the method signature: try `List(1,2,3).part<TAB>`, which completes to `List(1,2,3).partition`; press tab again to display `def partition(p: Int => Boolean): (List[Int], List[Int])`
* It even completes bean getters without typing `get`: try `(d: java.util.Date).day<TAB>`!

Review by @som-snytt

### CODE REVIEW

**Comment by retronym** [comment]
These commits, slated for 2.12, remove the old implementation and experiment with colorized completion suggestions: https://github.com/scala/scala/compare/f279894a3efe84b85e9becfd0ce96aaa0c21bfbd...d05b097e8abd56ddec1e22471a25de122ffd03c7

We should also upgrade JLIne 2.13.1 when it is ready (https://github.com/scala/scala-dev/issues/44)

**Comment by retronym** [comment]
Looks like I've missed the `any2stringAdd` implicit shadowing commit in my rebase. Mabye that can come in a separate PR, assuming it isn't dependent on these changes.

**Comment by som-snytt** [comment]
Sorry I've been squeezed for time by life + work, the classic rock + hard place respectively.

I think I included that SI-1931 commit b/c what it touches and likely conflicts, ugh. Including test output. I tried & failed one day to normalize the line numbering; I didn't try hard.

I see some of the previous commits were squashed? I'm not partial to history that way, but hard to track at a distance. You've got to raise the barrel a bit and account for wind, then maybe you hit the target? etc.

Soon we can use "resugar" without quotes. I mean resugar.

I hope someday you can comment on an ammonite issue, "fixed in scala repl." Which I guess needs a name now, someone should create an issue.

File: src/repl/scala/tools/nsc/interpreter/MemberHandlers.scala [hunk 1]
```diff
@@ -35,7 +35,18 @@ trait MemberHandlers {
...
+        else {
+          importVars += name
+          if (name.endsWith(IMain.DummyCursorFragment)) {
+            val stripped = name.stripSuffix(IMain.DummyCursorFragment)
+            importVars += stripped
+            if (stripped.isTypeName)
+            // Needed to import `xxx` during line 2 of:
+            //   scala> val xxx = ""
+            //   scala> def foo: x<TAB>
+            importVars += stripped.toTermName
```
Comment (janekdb) [inline src/repl/scala/tools/nsc/interpreter/MemberHandlers.scala]: Indent?

Comment (som-snytt) [inline src/repl/scala/tools/nsc/interpreter/MemberHandlers.scala]: The comment has to precede the if statement. I noticed it, too, but hoped the typesetter would catch it.

Comment (retronym) [inline src/repl/scala/tools/nsc/interpreter/MemberHandlers.scala]: Fixed. The last two lines were actually not needed.

**Comment by retronym** [comment]
@som-snytt I usually err on the side of submitting a history of commits that tells the story sub-feature-wise. In this case, given that the actual history was so meandering, and that all of the code here is new, I was a bit more squashy than usual.

**Comment by retronym** [comment]
Oh, and I'm pretty sure that ammonite (and Scala IDE) will need to play catch up with `global.def<TAB>`.

**Comment by retronym** [comment]
Added a few commits as a result of @SethTisue's feedback; results are sorted, and we hide inherited members of the interpreter wrappers.

**Comment by retronym** [comment]
I'm not going to include this in this PR, but I couldn't resist trying out some of this hole driven programming that the cool kids are raving about:

https://github.com/retronym/scala/compare/topic/completely-2.11...retronym:topic/completely-expected-type?expand=1

```
scala> class A; object A extends A; def foo(a: A) = 0; foo(__
   expected type: A

   ???   A   error

scala> class A; object A extends A; def foo(a: Int) = 0; foo(__)
   expected type: Int

   ##   ???   Integer2int   error   foo   hashCode   readInt
```

**Comment by retronym** [comment]
Another teaser: we can hijack keybindings in `jline.InteractiveReader` as follows:

```
        val CTRL_Q = 17.toChar
        this.getKeys.bind(CTRL_Q.toString, new ActionListener {
          override def actionPerformed(e: ActionEvent): Unit = {
            println("CTRL-Q")
            redrawLineAndFlush()
          }
        })
```

Could use this for a more convenient version of `//print`, or to display type errors before submitting the current line, or ...

**Comment by retronym** [comment]
I've added a commit with a cleaned up version of the camel case completion.

**Comment by SethTisue** [comment]
the camel case stuff is great, and worked on everything I tried it on

**Comment by som-snytt** [comment]
This isn't your fault, but right arrow sorts to the bottom. VBAR too. As opposed to category symbols coming first. Because we love symbols and encourage their use.

OK, I see how the getCamel thing works, neat.

I kind of think this should work: `classOf[String].typename[\t]`.  Another common one would be anything with `classpath` in it.

Basically, all my woes are due to caps; I was going to add ignorecase eventually. I have one machine where it's enabled in shell, another not because I'm lazy to look up the syntax, so I suffer. Yes, I suffer, but I appreciate it all the more when it works.

So I'd want not only `typ` to complete b/c `getTyp`, but also `all` because it's on a camel boundary in `notifyAll`.

I guess we'd want `mous` to complete to `getMouse` but not `isAnonymous`?

You probably noticed it offers `<byname>` et al.

I don't suppose you'll be offering XML completion? `:load mydoc.dtd`  Just kidding.

The new name for Typesafe could be `403 possibilities`.

`scala> Display all 403 possibilities? (y or n)`

It's biased to beans: `inst` -> `isInstanceOf` despite `asInstanceOf`.

More reasonably `buffered` trumps `toBuffer` on iterator, or `indexOf/Where` for `toIndexedSeq`.

But should `to` completes include `copyToArray`? `padTo`?

**Comment by som-snytt** [comment]
@SethTisue What about chillblains? Does it work on chillblains?

**Comment by som-snytt** [comment]
I don't need a support subscription to start using this as my regular REPL, do I?

**Comment by retronym** [comment]
> I don't need a support subscription to start using this as my regular REPL, do I?

Are you using it in prod? Or to prod?

**Comment by retronym** [comment]
> You probably noticed it offers <byname> et al.

I just put them back in their box.

**Comment by retronym** [comment]
```
[ERROR] Internal error: java.lang.RuntimeException: Failed to load p2 repository with ID 'scala-refactoring' from location file:/home/jenkins/workspace/scala-2.11.x-integrate-ide/target/scala-refactoring/org.scala-refactoring.update-site/target/site: No repository found at file:/home/jenkins/workspace/scala-2.11.x-integrate-ide/target/scala-refactoring/org.scala-refactoring.update-site/target/site. -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: Failed to load p2 repository with ID 'scala-refactoring' from location file:/home/jenkins/workspace/scala-2.11.x-integrate-ide/target/scala-refactoring/org.scala-refactoring.update-site/target/site
```

Is the IDE build failing out of jealousy?

**Comment by SethTisue** [comment]
tracking the integrate-ide failure issue at https://github.com/scala-ide/uber-build/issues/82

**Comment by retronym** [comment]
> Basically, all my woes are due to caps;

The last commit will save your ⇧ key.

**Comment by retronym** [comment]
One more commit to hide `asInstanceOf`, `ensuring` et al on the first ⇥ with no prefix.

**Comment by kiritsuku** [comment]
Bug with backticks:

```
scala> object O { def `a b` = 0 }
defined object O

// O.<tab> results in:
scala> O.a b
<console>:13: error: value a is not a member of object O
       O.a b
         ^
```

**Comment by som-snytt** [comment]
Backticks are the worst bugs because they carry Lyme disease.

https://issues.scala-lang.org/browse/SI-6919

https://issues.scala-lang.org/browse/SI-7957

**Comment by retronym** [comment]
@sschaef I've just pushed an additional commit that workaround the JLine bug, so completion at the second-to-last character of the buffer should work now.

**Comment by retronym** [comment]
I've noticed that completion of annotations is currently broken. This is a general problem with the presentation compiler.

 I've sketched out a fix in the last commit here: https://github.com/scala/scala/compare/scala:2.11.x...retronym:topic/ximporter-annotation-completion?expand=1, but it is a bit too involved to tack on to this already overweight PR.

**Comment by som-snytt** [comment]
Also fixes https://issues.scala-lang.org/browse/SI-7978. It would be handy if partest handled pending tests so that merging moved pending tests to active and then transitioned the related issue.

**Comment by retronym** [comment]
I've dropped the `classOf` changes for now and tracked them as https://issues.scala-lang.org/browse/SI-9477.

**Comment by lrytz** [comment]
Just saw @lihaoyi's discussion about memory usage over at https://github.com/lihaoyi/Ammonite/issues/216. It might be worth taking a look at the impact on memory usage of this PR.

**Comment by retronym** [comment]
Regarding memory, we create and discard the presentation compiler at each press of TAB. This is in addition to the resident compiler that is held for the duration of the REPL session. I think this is fairly low risk: for applications that are embedding the interpreter and that might be memory sensitive, no cost is incurred (they don't trigger autocompletion). For interactive users with JLine, the short lifetime should avoid the extra objects surviving to the old generation.

**Comment by retronym** [comment]
Heap usage looks okay to me. I was manually testing, autocompleting collections and the like:

image

**Comment by lihaoyi** [comment]
FWIW my memory usage conclusion wasn't wrt total memory held, but memory retention behavior. You guys set a max of 256mb somewhere in your she'll script, which works great, but I wanted to leave Ammonite's open ended because I've used it for some relatively large non-memory small-dara work.

The interesting conclusion is that G1 keeps the heap smallish when the working set is smallish, whereas the default GC let's it grow to the max heap size, e.g. 4gb on my laptop, even if the working set never goes above 100-200mb.

Ammonite keeps the pressy around but initializes it lazily, so it isn't part of that trial either =)

**Comment by lrytz** [comment]
thanks for checking the heap usage. i went through all the commits, it looks excellent. like everybody else, i can't wait to have it in :)

**Comment by lrytz** [comment]
LGTM

**Comment by lrytz** [comment]
/synch - maybe that helps with the cla status? -- yep it did -- but it also made scala-jenkins remove the "reviewed" label -- and it's quite stubborn about it.

**Comment by retronym** [comment]
@lrytz I've removed the unused parameter you flagged and made another minor refactoring in the last commit.

**Comment by lrytz** [comment]
LGTM - Let's Get This Merged :)

**Comment by lrytz** [comment]
:cake: !

**Comment by som-snytt** [comment]
:sweat_drops: what an effort!

:camel: crosses the REPL waste and also handles :camel: case.

:clock1130: is when he was up checking the heap.

**Comment by som-snytt** [comment]
The commit for SI-1931 went in under https://github.com/scala/scala/pull/4554

File: src/interactive/scala/tools/nsc/interactive/CompilerControl.scala [hunk 2]
```diff
@@ -288,6 +290,8 @@ trait CompilerControl { self: Global =>
     accessible: Boolean,
     inherited: Boolean,
     viaView: Symbol) extends Member {
+    // should be a case class parameter, but added as a var instead to preserve compatibility with the IDE
+    var prefix: Type = NoType
```
Comment (dragos) [inline src/interactive/scala/tools/nsc/interactive/CompilerControl.scala:294]: I think source-compatibility should be ok... Do we pattern match on this? :)

Comment (retronym) [inline src/interactive/scala/tools/nsc/interactive/CompilerControl.scala:294]: Yeah, I think you do. We'd need to phase that out first

**Pull request merged**